File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.cpp
// Search in Rotated Sorted Array
#include <vector>
using namespace std;
class Solution {
    int find_min(vector<int> &nums) {
        int left = 0, right = nums.size() - 1;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] <= nums[nums.size() - 1]) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

  public:
    int search(vector<int> &nums, int target) {
        int min_idx = find_min(nums);
        int left, right;
        if (target <= nums[nums.size() - 1]) {
            left = min_idx, right = nums.size() - 1;
        } else {
            left = 0, right = min_idx - 1;
        }
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) return mid;
            else if (nums[mid] > target) right = mid - 1;
            else left = mid + 1;
        }
        return -1;
    }
};

File: Data Structures & Algorithms/two-integer-sum/submission-4.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_dict = dict()
        for i in range(len(nums)):
            num_dict[nums[i]] = i

        for i, x in enumerate(nums):
            if target - x in num_dict and num_dict[target - x] != i:
                return [i, num_dict[target - x]]
        return []